framework/src/tinyalsa/tinyalsa.c

```diff
@@ -532,50 +533,82 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
 #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
 #else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
 #endif
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
 	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
-		return -errno;
-	} else {
-		pcm->running = 1;
+	if (!pcm->running) {
+		pcm->enqueuedBytes = 0;
+		/* If the device is opened for read and Our buffers are not enqued. Enque them now. */
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+		bufdesc.session = pcm->session;
+#endif
+#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
+		bufdesc.numbytes = pcm->config.period_size;
+		for (pcm->bufPtr = 0; pcm->bufPtr < buf_info.nbuffers; pcm->bufPtr++)
+#else
+		bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
+		for (pcm->bufPtr = 0; pcm->bufPtr < CONFIG_AUDIO_NUM_BUFFERS; pcm->bufPtr++)
+#endif
+		{
+			bufdesc.u.pBuffer = pcm->pBuffers[pcm->bufPtr];
+			if (nbytes <= pcm->enqueuedBytes) {
+				break;
+			} else if (nbytes - pcm->enqueuedBytes < bufdesc.numbytes) {
+				bufdesc.numbytes = nbytes - pcm->enqueuedBytes;
+			}
+			bufdesc.u.pBuffer->nmaxbytes = bufdesc.numbytes;
```

Comment:
If user requested 100 bytes in first readi call, and we have 4 buffers of 4096 bytes.
You would be restricting all 4 buffers to 100 bytes. So, there is under utilization of buffers. 
If user tries to read 4096 bytes in next readi, he will have to wait for multiple buffers.

Reply:
This is being done only for this buffer being enqueued, the value by default is the value set by config and determined at lines 551:555. The less number of bytes is sent only if less number of bytes are needed , I would not enqueue 4 buffers each of 100 bytes, instead just one buffer with nmaxbytes set to 100 bytes will be enqueued.